### ISSUE TITLE
Passing ADO.NET provider invariant name to RelationalStorage

### ISSUE THREAD

**Issue by veikkoeeva** [dotnet/orleans#569]
Currently there is no method to provide a ADO.NET provider invariant name to [RelationalStorage](https://github.com/dotnet/orleans/blob/75b93122a4f539f308b54e400c91ac4ed39f87b9/src/Orleans/Storage/RelationalStorage.cs#L150) and consequently the needed piece of data is [hardcoded like this](https://github.com/dotnet/orleans/blob/75b93122a4f539f308b54e400c91ac4ed39f87b9/src/Orleans/Messaging/SqlMembershipTable.cs#L50).

The invariant name is needed to load the correct database connector dlls as explained in the official MSDN documentation [Obtaining a DbProviderFactory](https://msdn.microsoft.com/en-us/library/dd0w4a2z%28v=vs.110%29.aspx) or as explained concisely in the Teradata blog [DbProviderFactories Demystified](https://developer.teradata.com/blog/netfx/2010/12/dbproviderfactories-demystified). The connection strings Orleans uses are defined by the special ``SystemStoreType`` element, as explained at [SQL System Storage](https://dotnet.github.io/orleans/Orleans-Configuration-Guide/Server-Configuration) which do not have a method to pass the invariant name.

In order to keep the connection string and invariant name in same place, which would feel logical, one option would be to add a new attribute to ``SystemStoreType`` that would tell the invariant name and pass that information in the configuration objects down to the respective runtime objects. Then in the ``App.config``/``Web.Config`` one could define the ``DbProviderFactories`` elements as usual with any other type of application. Other two options (are there more?) would be to either mandate the dlls need always to be intalled in GAC and added to ``machine.config`` where they are found or the config ``DbProviderFactory`` uses altered via runtime by reflection.

Having only GAC and ``machine.config`` would naturally make use of Nuget packages (e.g. [MySQL](https://www.nuget.org/packages/MySql.Data/)) more difficult and deployment and proof-of-concept which make development scenario more cumbersome, especially in enterprise scenarios where approval and installation process are more formal.

**&lt;edit:** I didn't elaborate on why it would feel logical to have the provider name in the same place as the connection string. The reason is that if this were a connection string defined in [connectionStrings](https://msdn.microsoft.com/en-us/library/ms254494(v=vs.110).aspx) block in ``.config`` file, it would be given there in the ``providerName`` element. By default it is ``System.Data.SqlClient``.

Also, it would be possible to embed ``providerName`` to the connection string itself and strip it off to be used normally before using connection string.

**Comment by jthelin** [comment]
Please, lets avoid hard requiring pre-install into GAC at all costs!

Xcopy onto baseline machine images is the new norm for deployment to large number of cluster, even preferred over scriptable component pre-installs.

**Comment by sergeybykov** [comment]
+1 for no GAC and for not breaking the xcopy style deployment.

**Comment by veikkoeeva** [comment]
A few more ideas about possible implementation:

1. Have it as a separate key-value pair in the regular .config file. Though I still believe the best would be to have it together with the connection string.
2. Alter the configuration (also [OrleansConfiguration.xsd](https://github.com/dotnet/orleans/blob/master/src/Orleans/Configuration/OrleansConfiguration.xsd)) so that ``SystemStoreType`` would be either one of the provider invariant names or, say, ``AzureTable`` in case of Azure Table Storage.

**Comment by shayhatsor** [comment]
I want to implement a MySQL reminders provider.
after reading the docs on relational storage, this issue and #648, I think I figured out what's needed to do that:
1) prepare a MySQL DB that contains the Reminders related tables, and queries
2) copy paste [SqlMembershipTable.cs](../blob/master/src/Orleans/Messaging/SqlMembershipTable.cs) to a new project
3) replace [This Line](../blob/master/src/Orleans/Messaging/SqlMembershipTable.cs#L79) with 
```c#
database = RelationalStorageUtilities.CreateGenericStorageInstance("MySql.Data.MySqlClient", config.DataConnectionString); 
```
4) add the assembly as explained in #648

the problem is that some of the methods on [OrleansRelationalExtensions.cs](../blob/master/src/Orleans/RelationalStorage/OrleansRelationalExtensions.cs) are **internal**.

The thing is that this is a workaround for the current issue, I'd rather just fix this issue properly.
should i just add a key to the configuration that has a default value of **System.Data.SqlClient** ?
The only problem with this is that the same InvariantName will be used for **All** providers, which will not allow ZooKeeper for membership and MySQL for reminders as they read from the same connection string property. What I'm thinking is that the Orleans services : Membership, Reminders and Statistics interfaces should extend **IProvider**. And in the global settings there would be an IProviderConfiguration for each of the services.
what do you think ? @gabikliot and @veikkoeeva

**Comment by veikkoeeva** [comment]
@shayhatsor 

1) That is right. You need only a subset of the queries, but it might be better to provide all the keys (with dummy queries?). I'm currently refactoring (see [here](https://github.com/dotnet/orleans/pull/735) currently) it a bit and I'm thinking to add a check that the number of queries read from the database upon initialization matches that of expected.

2) Reminders are implemented in [SqlRemindersTable](https://github.com/dotnet/orleans/blob/master/src/OrleansRuntime/ReminderService/SqlReminderTable.cs). The logic remains the same however (written to the database using hand tuned queries).

3) I think you would not need to make a one-off fork of this class if we had a method to pass the vendor invariant string in a way or another to these classes. The invariant name is needed so that ADO.NET knows which dlls to load, which in your case are likely the Nuget libraries. The invariant name is also added to a constant string as I wrote the original code to use a named constant instead of a string.

4) I think the best way would be to add a key to the configuration and pass that to everywhere the ADO.NET libraries are loaded and then it should work by just providing a suitable MySQL script. Currently they are the [SqlMembershipTable](https://github.com/dotnet/orleans/blob/master/src/Orleans/Messaging/SqlMembershipTable.cs), [SqlRemindersTable](https://github.com/dotnet/orleans/blob/master/src/OrleansRuntime/ReminderService/SqlReminderTable.cs) and [SqlStatisticsPublisher](https://github.com/dotnet/orleans/blob/master/src/OrleansProviders/SQLServer/SqlStatisticsPublisher.cs).

That some methods are internal is an oversight, but thinking of it now, there is no reason they should be public either.

With regard to new backends, there's a [SqlTestsEnvironment](https://github.com/dotnet/orleans/blob/master/src/TesterInternal/SqlTestsEnvironment.cs) that can be used to initialize a new database and test it automatically (in my refactoring branch I relocated it [here](https://github.com/dotnet/orleans/pull/735/files#diff-f4a0f0327de199c1910aa6c39e932f61R1) and refactored it a bit to make it easier to add new relational backends). The idea is that in tests one providers a few SQL clauses to create a database, initialize it with the same scripts as Orleans does and the tests would be automatically run. Currently the SQL Server version uses a default connection string, I think all databases have a "default settings", i.e. ports etc. they install themselves to unless otherwise instructed. However, this is not a fixes and the connection string can be provided by whatever means and all code refactored as needed (even so that it can run the code against all db's it can find and connect to and find a script to).

What else...
There are a few reminder tests too already and the SQL Server script has a lot of optimization etc., not all of that is needed. I try to finish refactoring soon, but it's all right if you decide go ahead already. Also, I'm hanging on the Orleans channel if you want to have a chat.

**Comment by gabikliot** [comment]
Closed via #750.
